src/top/utop.el

```diff
@@ -604,12 +604,10 @@ it is started."
          (dolist (prefix utop-completion-prefixes)
            ;; We need to handle specially prefixes like "List.m" as
            ;; the responses from utop don't include the module prefix.
-           (let ((prefix (if (string-match-p "\\." prefix)
-                             (cadr (split-string prefix "\\."))
-                           prefix)))
-             (when (string-prefix-p prefix argument)
-               (push argument utop-completion)
-               (throw 'done t))))))
+           (when-let* ((pos (string-match  "[^.]*$" prefix))
+                       ((string-prefix-p (substring prefix pos) argument)))
+             (push (concat (substring prefix 0 pos) argument) utop-completion)
```

Comment:
Probably a binding for `(substring prefix 0 pos)` will make this bit of code slightly easier to understand.

Reply:
Thanks for your helpful suggestions. I amended the commit.